airgun/widgets.py

```diff
@@ -1267,6 +1267,89 @@ def read(self):
+    """
+    ROOT = "//form[contains(@class, 'content-view-pf-pagination')]"
+    # Kattelo views use per_page with select, foreman use a per_page with Button DropDown.
+    PER_PAGE_BUTTON_DROPDOWN = ".//div[button[@id='pagination-row-dropdown']]"
+    PER_PAGE_SELECT = ".//select[contains(@ng-model, 'per_page')]"
+    first_page_button = Text(".//li[a[span[contains(@class, 'angle-double-left')]]]")
+    previous_page_button = Text(".//li[a[span[contains(@class, 'angle-left')]]]")
+    next_page_button = Text(".//li[a[span[contains(@class, 'angle-right')]]]")
+    last_page_button = Text(".//li[a[span[contains(@class, 'angle-double-right')]]]")
+    page = TextInput(locator=".//input[contains(@class, 'pagination-pf-page')]")
+    pages = Text(".//span[contains(@class, 'pagination-pf-pages')]")
+
+    def _element_exists(self, locator, parent=None, check_visibility=False):
+        """Check if a locator element/widget exists in the DOM"""
+        try:
+            exists = bool(self.browser.elements(
+                locator, parent=parent, check_visibility=check_visibility))
+        except NoSuchElementException:
+            exists = False
+        return exists
+
+    @cached_property
+    def per_page(self):
+        """Return the per page widget"""
+        if self._element_exists(self.PER_PAGE_SELECT, parent=self):
+            return Select(self, self.PER_PAGE_SELECT)
+        return ActionsDropdown(self, self.PER_PAGE_BUTTON_DROPDOWN)
+
+    @property
+    def is_displayed(self):
+        """Check whether this Pagination widget exists and visible"""
+        return self._element_exists(self.pages, parent=self, check_visibility=True)
+
+    def _click_button(self, pager_button):
+        """Click on the pager button if enabled."""
+        if "disabled" not in self.browser.classes(pager_button):
+            pager_button.click()
+        else:
+            raise DisabledWidgetError('Button {0} is not enabled'.format(pager_button))
+
+    def first_page(self):
+        """Goto first page by clicking on first page button"""
+        self._click_button(self.first_page_button)
+
+    def previous_page(self):
+        """Goto previous page by clicking on previous page button"""
+        self._click_button(self.previous_page_button)
+
+    def next_page(self):
+        """Goto next page by clicking on next page button"""
+        self._click_button(self.next_page_button)
+
+    def last_page(self):
+        """Goto last page by clicking on last page button"""
+        self._click_button(self.last_page_button)
+
+    @property
+    def current_page(self):
+        """Return the current page as integer value"""
+        return int(self.page.read())
+
+    @property
+    def total_pages(self):
+        """Return the total available pages as integer value"""
+        return int(self.pages.read())
+
+    def read(self):
+        """Read the more valuable sub widgets of this pagination widget"""
```

Comment:
not sure that I understand word 'valuable' in that context

Reply:
The needed ones, any way we can remove completely that doc, but we need that per_page widget value any way, as it is not returned if we go through this widget sub-widgets (by the usual way).

Reply:
do you know meaning of word 'valuable'? I just can't connect it to that context